### ISSUE TITLE
False positive attribute-defined-outside-init when setting variant field

### ISSUE THREAD

**Issue by anordal** [pylint-dev/pylint#8900]
### Bug description

```python
class MessageVariant:
    message: str

    def __init__(self, message: str) -> None:
        self.message = message


class OtherVariant:
    pass


def parse(operation: str) -> MessageVariant | OtherVariant:
    if operation != "":
        return MessageVariant(operation)
    return OtherVariant()


def edit_messages(operation: str) -> None:
    variant = parse(operation)
    if isinstance(variant, MessageVariant):
        # W0201: Attribute 'message' defined outside __init__ (attribute-defined-outside-init)
        variant.message = ""
```

### Command used

```shell
python3.10 -m pylint --disable=C0114,C0115,C0116,R0903 edit_messages.py
```

### Pylint output

```shell
edit_messages.py:23:8: W0201: Attribute 'message' defined outside __init__ (attribute-defined-outside-init)
```

### Expected behavior

As far as I can see, the message _is_ defined in __init__, so I don't expect a warning about that.

### Pylint version

```shell
python3.10 -m pylint --version
pylint 2.17.4
astroid 2.15.6
Python 3.10.12 (main, Jun 28 2023, 17:56:55) [GCC]
```

### OS / Environment

openSUSE Tumbleweed 20230728
